File: python/树/559. N叉树的最大深度.py
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children

class Solution:
    def maxDepth(self, root: 'Node') -> int:
        if not root:
            return 0
        from collections import deque
        queue = deque()
        queue.append(root)
        count = 0
        while queue:
            for _ in range(len(queue)):
                node = queue.popleft()
                if node.children:
                    for child in node.children:
                        queue.append(child)
            count+=1
        return count

# 递归
class Solution(object):
    def maxDepth(self, root):
        """
        :type root: Node
        :rtype: int
        """
        if root is None:
            return 0
        elif root.children == []:
            return 1
        else:
            height = [self.maxDepth(c) for c in root.children]
            return max(height) + 1

File: go/general/559. N叉树的最大深度.go
package main

import "math"

func maxDepth(root *Node) int {
	if root == nil {
		return 0
	}
	if len(root.Children) == 0 {
		return 1
	}
	var height []int
	for _, v := range root.Children {
		height = append(height, maxDepth(v))
	}
	return max(height) + 1
}

func max(h []int) int {
	_max := -1
	for _, v := range h {
		if v > _max {
			_max = v
		}
	}
	return _max
}
